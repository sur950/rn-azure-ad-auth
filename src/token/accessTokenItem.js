import Scope from './scope'
import BaseTokenItem from './baseTokenItem'

const DEFAULT_EXPIRE_IN_SECONDS = 3600
const DEFAULT_EXPIRATION_BUFFER = 300

/**
 * Class represent acces token cache item
 *
 * @namespace TokenCache.AccessTokenItem
 *
 * @param {Object} tokenResponse
 * @param {String} clientId
 *
 * @class AccessTokenItem
 * @extends BaseTokenItem
 * @memberof TokenCache
 */
export default class AccessTokenItem extends BaseTokenItem{
    constructor(tokenResponse, clientId) {
        if (!tokenResponse.accessToken) {
            throw new Error('Invalid token response. Can not create access token.')
        }
        super(tokenResponse, clientId)

        this.accessToken = tokenResponse.accessToken
        this.scope = typeof tokenResponse.scope != 'undefined' ? new Scope(tokenResponse.scope) : Scope.basicScope()
        this.expireOn = Date.now() + (tokenResponse.expiresIn ? parseInt(tokenResponse.expiresIn) : DEFAULT_EXPIRE_IN_SECONDS)*1000
    }

    tokenKey() {
        return BaseTokenItem.createAccessTokenKey(this.clientId, this.userId, this.scope)
    }

    isExpired() {
        const validity = Date.now() + DEFAULT_EXPIRATION_BUFFER
        return this.expireOn < validity
    }

    static fromJson(objStr) {
        const rawObject = BaseTokenItem.rawObjectFromJson(objStr)
        if (!rawObject || !rawObject['accessToken']) {
            return null
        }
        const obj = Object.create(AccessTokenItem.prototype)
        return Object.assign(obj, rawObject)
    }

}