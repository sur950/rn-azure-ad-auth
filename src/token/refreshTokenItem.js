import BaseTokenItem from './baseTokenItem'

/**
 * Class represent refresh token cache item
 *
 * @namespace TokenCache.RefreshTokenItem
 *
 * @param {Object} tokenResponse
 * @param {String} clientId
 *
 * @class RefreshTokenItem
 * @extends BaseTokenItem
 * @memberof TokenCache
 */
export default class RefreshTokenItem extends BaseTokenItem{
    constructor(tokenResponse, clientId) {
        if (!tokenResponse.refreshToken) {
            throw new Error('Invalid token response. Can not create refresh token.')
        }
        
        super(tokenResponse, clientId)
        this.refreshToken = tokenResponse.refreshToken
    }

    tokenKey() {
        return BaseTokenItem.createRefreshTokenKey(this.clientId, this.userId)
    }

    static fromJson(objStr) {
        const rawObject = BaseTokenItem.rawObjectFromJson(objStr)
        if (!rawObject || !rawObject['refreshToken']) {
            return null
        }
        const obj = Object.create(RefreshTokenItem.prototype)
        return Object.assign(obj, rawObject)
    }
    
}