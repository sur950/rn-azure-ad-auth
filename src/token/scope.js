const BASIC_SCOPE = 'offline_access openid profile'
/**
 * Azure AD Auth scope representation class
 *
 * 1. Remove MS Graph URLs from scope, as it is default for any scope
 * 2. Remove eventual commas and double spaces
 * 3. Sort
 * 4. BASIC SCOPE is always a part of auth requests
 *
 * @namespace TokenCache.Scope
 *
 * @param {string | Array<String> | ''} scope - without parameters represents
 * BASIC_SCOPE = 'offline_access openid profile'
 *
 * @memberof TokenCache
 * @class Scope
 */

export default class Scope {
    constructor(scope = '') {
        if(scope instanceof Scope) {
            this.scopeStr = scope.scopeStr
        } else if (typeof scope === 'string') {
            this.scopeStr = this._refine(scope)
        } else if (scope.constructor === Array) {
            this.scopeStr = this._refine(scope.join(' '))
        } else {
            throw new Error('Ivalid constructor parameter in Scope()')
        }

        if (this.scopeStr.length == 0) {
            this.basicScope = true
            this.scope = []
        } else {
            this.scope = this.scopeStr.split(' ')
        }

    }

    toString() {
        if (this.basicScope) {
            return BASIC_SCOPE
        }
        return BASIC_SCOPE + ' ' + this.scopeStr
    }

    _refine(scope) {
        // remove standard scopes, that are added to every request
        let newScope = scope.replace(/openid|profile|offline_access/g, '').toLowerCase()
        // remove MS Graph URL
        newScope = newScope.replace(new RegExp('https://graph.microsoft.com/', 'g'), '')
            // remove double spaces and commas
            .replace(/,+/g, '').replace(/\s+/g, ' ').trim()
        // Sort
        return newScope.split(' ').sort().join(' ')
    }
    
    /**
     * Check if newScope is a subset of baseScope
     *
     * @param {Array} newScope
     * @param {Array} otherScope
     *
     * @returns {boolean}
     */
    isSubsetOf(otherScope) {
        if (!(otherScope instanceof Scope)) {
            throw new Error('Parameter is not a Scope() instance')
        }
        if (otherScope.basicScope || this.basicScope) {
            // The condition order here is important!
            // With the second condition we should only return true if the first one is false
            return true
        }
        let difference = this.scope.filter(x => !otherScope.scope.includes(x))
        return difference.length == 0
    }
    
    /**
     * Copmare if the current instance scope intersects witt one from parameter
     * Only NON basic scopes are compared
     *
     * @param {Scope} otherScope
     */
    isIntersects(otherScope) {
        if (!(otherScope instanceof Scope)) {
            throw new Error('Parameter is not a Scope() instance')
        }

        for (const element of this.scope) {
            if (otherScope.scope.includes(element)) {
                return true
            }
        }
        
        return false
    }

    equals(other) {
        if (!(other instanceof Scope)) {
            throw new Error('Parameter is not a Scope() instance')
        }
        return this.toString() === other.toString()
    }

    static basicScope() {
        return new Scope()
    }
}