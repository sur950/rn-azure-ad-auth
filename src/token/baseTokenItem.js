import { extractIdToken } from './token'
import { Base64 } from 'js-base64'
import Scope from './scope'

const TOKEN_CACHE_KEY_DELIMITER = '$'

function normalizeId(id) {
    if (!id) {
        throw new Error('Id is null or undefined: ', id)
    }
    if (typeof id === 'string' || id instanceof String) {
        return id.toLocaleLowerCase()
    }
    if (typeof id === 'number') {
        return id.toString()
    }
    throw new Error('Invalid id: ', id)
}

/**
 * Class represent basic token cache item
 *
 * Note: userId is handled in case insencitive way for token cache keys
 *
 * @namespace TokenCache.BaseTokenItem
 *
 * @param {Object} tokenResponse
 * @param {String} clientId
 *
 * @class BaseTokenItem
 * @memberof TokenCache
 */

export default class BaseTokenItem {
    constructor(tokenResponse, clientId) {
        this.clientId = clientId
        this.rawIdToken = tokenResponse.idToken
        let decodedIdToken = extractIdToken(tokenResponse.idToken)

        this.userId = decodedIdToken.preferred_username || decodedIdToken.unique_name || decodedIdToken.sub

        this.userName = decodedIdToken.name
        if (!this.userName && decodedIdToken.given_name && decodedIdToken.family_name) {
            this.userName = decodedIdToken.given_name + ' ' + decodedIdToken.family_name
        } else if (!this.userName) {
            this.userName = this.userId
        }

        if (decodedIdToken.tid) {
            this.tenantId = decodedIdToken.tid
        } else if (decodedIdToken.iss) {
            const iss = decodedIdToken.iss
            const b2cSuffixIndex = iss.indexOf('.b2clogin.com')
            // parse then tenant ID out of the issuer claim
            this.tenantId = iss.substring('https://'.length, b2cSuffixIndex > 0 ? b2cSuffixIndex : iss.length)
        }
        this.idTokenExpireOn = parseInt(decodedIdToken.exp)*1000
    }

    static createRefreshTokenKey(clientId, userId) {
        return Base64.encodeURI(clientId) +
            TOKEN_CACHE_KEY_DELIMITER +
            Base64.encodeURI(normalizeId(userId))
    }

    static createAccessTokenKey(clientId, userId, scope) {
        return Base64.encodeURI(clientId) +
            TOKEN_CACHE_KEY_DELIMITER +
            Base64.encodeURI(normalizeId(userId)) +
            TOKEN_CACHE_KEY_DELIMITER +
            Base64.encodeURI(scope.toString())
    }

    static createTokenKeyPrefix(clientId, userId) {
        let prefix = Base64.encodeURI(clientId) + TOKEN_CACHE_KEY_DELIMITER
        if (userId) {
            prefix = prefix + Base64.encodeURI(normalizeId(userId))
        }

        return prefix
    }

    static scopeFromKey(key) {
        const keyParts = key.split(TOKEN_CACHE_KEY_DELIMITER)
        if (keyParts[2]) {
            const scopeStr = Base64.decode(keyParts[2])
            return new Scope(scopeStr)
        }
        return null
    }

    toString() {
        return JSON.stringify(this)
    }

    static rawObjectFromJson(objStr) {
        if (typeof objStr !== 'string' && !(objStr instanceof String) || objStr.length < 5) {
            return null
        }
        try {
            return JSON.parse(objStr)
        } catch (e) {
            console.warn('Incorect JSON token string: ', objStr)
            return null
        }
    }
}