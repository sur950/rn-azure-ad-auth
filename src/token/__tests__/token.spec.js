import { extractIdToken } from '../token'
import RefreshTokenItem from '../refreshTokenItem'
import AccessTokenItem from '../accessTokenItem'

const testIdToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IlNTUWRoSTFjS3ZoUUVEU0p4RTJnR1lzNDBRMCJ9.\
eyJhdWQiOiJiNWQxMjBmNi0wNGZiLTQ4MWEtODhmNi1iYzJkZmQ0NGZkYTciLCJpc3MiOiJodHRwczovL2xvZ2luLm1pY3Jvc29mdG9ubG\
luZS5jb20vMzFmMWI3ODktOTBlMy00NDJhLWFjZDItZDZhZThjOGJkYTMxL3YyLjAiLCJpYXQiOjE1MjA1MjQ0NjEsIm5iZiI6MTUyMDUy\
NDQ2MSwiZXhwIjoxNTIwNTI4MzYxLCJhaW8iOiJBVFFBeS84R0FBQUFWRnBvOVJjNENsbW1yUGFacXp4TTkvZ2JGb2hmOEZOZC9wNk14Sk\
Y1SUw5OU9sR0REeTFZcytQeWl6RGk4Y3NtIiwibmFtZSI6IlZsYWRpbWlyIE11cmluIChFeHRlcm4pIiwibm9uY2UiOiJZOWF5YjZoUEpL\
b2ZnQXZMU2xtOUxVVTdOQ1dmMVU2ZFZrdlZwVHRzclBFIiwib2lkIjoiZDQ1OTc4ZjAtOTVmMy00NGQyLWJiOWUtY2U5ODZjM2VkMjc1Ii\
wicHJlZmVycmVkX3VzZXJuYW1lIjoidmxhZGltaXIubXVyaW4uZXh0ZXJuQHBvcnNjaGUuZGlnaXRhbCIsInN1YiI6ImhCa3ZVUmNXZnNy\
SGxzQ2otQkFWcGJTTURPWVFfc2VXTDNuTFQ1RWRJanMiLCJ0aWQiOiIzMWYxYjc4OS05MGUzLTQ0MmEtYWNkMi1kNmFlOGM4YmRhMzEiLCJ\
1dGkiOiI0ejdHd2pQbnNVQ01Bc3VyNzRVTUFBIiwidmVyIjoiMi4wIn0.Uj7bD5SCNpJeQ73dcxi-wijwB2Zld_TudHk3FTYi6jVItobJiA\
xlTeEsFo9NNo14ls_z5_dGA8_O_SIy0OarvmsuicBWQ9N5DAG3Jwxxci399yx4sQwxJko0YoRx7W1Y2H8DeukE66DcQ8hJ3yZzX1xwFyrZ_\
fyzfG5SuIN-mUF9JZN5AvIJSdKJ7SiKKwGUNuz_ytHAumbyIzmv9T0922QdIL7bUB-KWdpo7dfM7wWD7mdGil9IB8_F-ovS2dvTnJafTO5s\
Xi92arsI9e05zNaGr0oEUZknwAI6SOcVXWZlN7D6RSH-vnOMidjHeQvGC3A9MOmrPs3-qEKvo6HVNQ'

describe('token actions', () => {
    it('should correctly extract Id token', () => {
        expect(extractIdToken(testIdToken)).toMatchSnapshot()
    })

    it('should contain user name', () => {
        expect(extractIdToken(testIdToken)).toHaveProperty('preferred_username')
    })
})

describe('refresh token', () => {
    let json = JSON.stringify(extractIdToken(testIdToken))
    it('should return null for empty JSON string input', () => {
        expect(RefreshTokenItem.fromJson('')).toBeNull()
    })
    it('should return null for incorrect JSON string input', () => {
        expect(RefreshTokenItem.fromJson('ssssss')).toBeNull()
    })
    it('should return null for invalid JSON string', () => {
        expect(RefreshTokenItem.fromJson(json)).toBeNull()
    })
})

describe('access token', () => {
    let json = JSON.stringify(extractIdToken(testIdToken))
    it('should return null for empty JSON string input', () => {
        expect(AccessTokenItem.fromJson('')).toBeNull()
    })
    it('should return null for incorrect JSON string input', () => {
        expect(AccessTokenItem.fromJson('ssssss')).toBeNull()
    })
    it('should return null for invalid JSON token', () => {
        expect(AccessTokenItem.fromJson(json)).toBeNull()
    })
})