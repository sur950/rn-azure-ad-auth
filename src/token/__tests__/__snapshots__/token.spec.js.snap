// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`token actions should correctly extract Id token 1`] = `
Object {
  "aio": "ATQAy/8GAAAAVFpo9Rc4ClmmrPaZqzxM9/gbFohf8FNd/p6MxJF5IL99OlGDDy1Ys+PyizDi8csm",
  "aud": "b5d120f6-04fb-481a-88f6-bc2dfd44fda7",
  "exp": 1520528361,
  "iat": 1520524461,
  "iss": "https://login.microsoftonline.com/31f1b789-90e3-442a-acd2-d6ae8c8bda31/v2.0",
  "name": "Vladimir Murin (Extern)",
  "nbf": 1520524461,
  "nonce": "Y9ayb6hPJKofgAvLSlm9LUU7NCWf1U6dVkvVpTtsrPE",
  "oid": "d45978f0-95f3-44d2-bb9e-ce986c3ed275",
  "preferred_username": "vladimir.murin.extern@porsche.digital",
  "sub": "hBkvURcWfsrHlsCj-BAVpbSMDOYQ_seWL3nLT5EdIjs",
  "tid": "31f1b789-90e3-442a-acd2-d6ae8c8bda31",
  "uti": "4z7GwjPnsUCMAsur74UMAA",
  "ver": "2.0",
}
`;
